### PR TITLE
Add secret repository metrics. Fixes #3942

### LINKED ISSUE #3942: Add secret repository metrics
We want to have better visibility over the time spent communicating with the various secret provider implementations. We should have metrics for all secret management operations.

### PR DESCRIPTION
**Pull request by alrod**
Resolves #3942

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Controllers/KeysController.cs [hunk 1]
```diff
@@ -28,17 +29,38 @@ public class KeysController : Controller
         private readonly ILogger _logger;
         private readonly IOptions<ScriptApplicationHostOptions> _applicationOptions;
         private readonly IFileSystem _fileSystem;
+        private readonly IMetricsLogger _metrics;
+        private readonly string _metricsDescription;
 
-        public KeysController(IOptions<ScriptApplicationHostOptions> applicationOptions, ISecretManagerProvider secretManagerProvider, ILoggerFactory loggerFactory, IFileSystem fileSystem)
+        public KeysController(IOptions<ScriptApplicationHostOptions> applicationOptions, ISecretManagerProvider secretManagerProvider, ILoggerFactory loggerFactory, IFileSystem fileSystem, IMetricsLogger metrics)
```
Comment (brettsam) [inline src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: I don't really like the requirement that to use `[MetricsAction]` you also need to implement `IMetricsController` (and have `IMetricsLogger` injected).

The `IMetricsLogger` requirement can be removed if you do something similar to the `RequiresRunningHostAttribute` -- where you can inject services directly to the attribute and the controller doesn't have to know about it. I'd recommend doing that. 

That leaves us with the `MetricsDescription`... trying to think of a way to remove that requirement.

File: src/WebJobs.Script.WebHost/Controllers/MertricsActionAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        public override void OnActionExecuting(ActionExecutingContext filterContext)
+        {
+            var controllerName = filterContext.RouteData.Values["Controller"];
+            var actionName = filterContext.RouteData.Values["Action"];
+
+            _controller = filterContext.Controller as IMetricsController;
+            if (_controller != null)
+            {
+                _metricEvent = _controller.MetricsLogger.BeginEvent($"{_controller.MetricsDescription}\\{controllerName}\\{actionName}");
+            }
```
Comment (brettsam) [inline src/WebJobs.Script.WebHost/Controllers/MertricsActionAttribute.cs]: Could we make implementing `IMetricsController` optional? Other controllers may just want to use this attribute to get metrics for their actions, but don't necessarily need to inject any additional data.

File: src/WebJobs.Script.WebHost/Controllers/MertricsActionAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Azure.WebJobs.Script.Diagnostics;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Controllers
+{
+    public class MertricsActionAttribute : ActionFilterAttribute
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Controllers/MertricsActionAttribute.cs]: Spelling (`MetricsActionAttribute`)

File: src/WebJobs.Script.WebHost/Controllers/KeysController.cs (hunk 1 above)
Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: Agreed. We already have a pattern established and that would be a cleaner approach for those cross cutting features. For the description, though it my be repetitive, to reduce the coupling, I would rather see that passed into the attribute as an argument.

Comment (brettsam) [reply src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: In this case, he needs the description to be dynamic, based on the repository type being used...

Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: I noticed that, but not sure what the exact requirements are. Do we need this to be repository metrics? If so, I'd question adding this to the controller to begin with.

Comment (alrod) [reply src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: As we do not have good way to pass `MetricsDescription` into the attribute I sent out an another PR  that adds metrics directly in `SecretManager ` https://github.com/Azure/azure-functions-host/pull/4122

**Pull request closed without merging**